src/v/kafka/server/handlers/produce.cc

```diff
@@ -257,31 +258,43 @@ reader_from_lcore_batch(model::record_batch&& batch) {
 static ss::future<produce_response::partition> partition_append(
   model::partition_id id,
   ss::lw_shared_ptr<cluster::partition> partition,
+  raft::batch_identity bid,
   model::record_batch_reader reader,
   int16_t acks,
   int32_t num_records) {
     return partition
-      ->replicate(std::move(reader), acks_to_replicate_options(acks))
-      .then_wrapped([partition, id, num_records = num_records](
-                      ss::future<result<raft::replicate_result>> f) {
-          produce_response::partition p{.id = id};
-          try {
-              auto r = f.get0();
-              if (r) {
-                  // have to subtract num_of_records - 1 as base_offset
-                  // is inclusive
-                  p.base_offset = model::offset(
-                    r.value().last_offset() - (num_records - 1));
-                  p.error = error_code::none;
-                  partition->probe().add_records_produced(num_records);
-              } else {
-                  p.error = error_code::unknown_server_error;
-              }
-          } catch (...) {
-              p.error = error_code::unknown_server_error;
-          }
-          return p;
-      });
+      ->replicate(bid, std::move(reader), acks_to_replicate_options(acks))
+      .then_wrapped(
+        [partition, id, num_records = num_records](
+          ss::future<checked<raft::replicate_result, kafka::error_code>> f) {
+            produce_response::partition p{.id = id};
+            try {
+                auto r = f.get0();
+                if (r.has_value()) {
+                    // have to subtract num_of_records - 1 as base_offset
+                    // is inclusive
+                    p.base_offset = model::offset(
+                      r.value().last_offset() - (num_records - 1));
+                    p.error = error_code::none;
+                    partition->probe().add_records_produced(num_records);
+                } else if (
+                  r
```

Comment:
i think it is time to split this function to deal with the indentation

Reply:
I've extracted long ad hoc object creation (`not_leader_for_partition(error_code::not_leader_for_partition)`) into consts to deal with the indentation